File: ts/src/modules/lexer/lexer.ts
import { TokenType, Tokens, Token, LookupIdent } from '../token'

type Char = string | number

export class Lexer {
  private position: number = 0;
  private readPosition: number = 0;
  private ch: Char = "";

  constructor(
    private input: string,
  ) {
    this.readChar()
  }

  public readChar() {
    if (this.readPosition > this.input.length) {
      this.ch = 0
    } else {
      this.ch = this.input[this.readPosition]! // guaranteed to be here
    }
    this.position = this.readPosition;
    this.readPosition++
  }

  public readIdentifier(): string {
    const start = this.position;
    while (isLetter(this.ch)) {
      this.readChar()
    }

    return this.input.substring(start, this.position)
  }

  public readNumber(): { literal: string, type: TokenType } {
    const start = this.position;
    let hasDot = false;
    while (isDigit(this.ch) || this.ch == '.') {
      if (this.ch == '.' && hasDot) {
        return { literal: "", type: Tokens.ILLEGAL }
      } else if (this.ch == '.') {
        hasDot = true
      }
      this.readChar()
    }
    if (hasDot) {
      return { literal: this.input.substring(start, this.position), type: Tokens.FLOAT }
    }
    return { literal: this.input.substring(start, this.position), type: Tokens.INT }
  }

  public skipWhitespace() {
    while (this.ch == ' ' || this.ch == '\t' || this.ch == '\n' || this.ch == '\r') {
      this.readChar()
    }
  }

  public peekChar(): Char {
    if (this.readPosition >= this.input.length) {
      return 0
    } else {
      return this.input[this.readPosition]!
    }
  }

  public nextToken(): Token {
    let tok: Token = { type: Tokens.EOF, literal: "" }

    this.skipWhitespace()

    switch (this.ch) {
      case '=':
        // if this.peekChar() == '=' {
        //   tok = l.newTwoCharToken(Tokens.EQ)
        // } else {
        tok = newToken(Tokens.ASSIGN, this.ch)
        // }
        break;
      case '-':
        tok = newToken(Tokens.MINUS, this.ch)
        break;
      case '!':
        // if l.peekChar() == '=' {
        //   tok = l.newTwoCharToken(Tokens.NOT_EQ)
        // } else {
        tok = newToken(Tokens.BANG, this.ch)
        // }
        break;
      case '/':
        tok = newToken(Tokens.SLASH, this.ch)
        break;
      case '*':
        tok = newToken(Tokens.ASTERISK, this.ch)
        break;
      case '<':
        tok = newToken(Tokens.LT, this.ch)
        break;
      case '>':
        tok = newToken(Tokens.GT, this.ch)
        break;
      case ';':
        tok = newToken(Tokens.SEMICOLON, this.ch)
        break;
      case '(':
        tok = newToken(Tokens.LPAREN, this.ch)
        break;
      case ')':
        tok = newToken(Tokens.RPAREN, this.ch)
        break;
      case ',':
        tok = newToken(Tokens.COMMA, this.ch)
        break;
      case '+':
        tok = newToken(Tokens.PLUS, this.ch)
        break;
      case '{':
        tok = newToken(Tokens.LBRACE, this.ch)
        break;
      case '}':
        tok = newToken(Tokens.RBRACE, this.ch)
        break;
      case 0:
        tok.literal = ""
        tok.type = Tokens.EOF
        break;
      default:
        if (isLetter(this.ch)) {
          tok.literal = this.readIdentifier()
          tok.type = LookupIdent(tok.literal)
          return tok
        } else if (isDigit(this.ch)) {
          tok = this.readNumber()
          return tok
        } else {
          tok = newToken(Tokens.ILLEGAL, this.ch)
        }

    }

    this.readChar()
    return tok

  }
}

function newToken(tokenType: TokenType, ch: Char): Token {
  return { type: tokenType, literal: String(ch) }
}

function isLetter(ch: Char): boolean {
  return !!ch && ('a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_')
}

function isDigit(ch: Char): boolean {
  return !!ch && '0' <= ch && ch <= '9'
}

File: rust/src/lexer.rs
use anyhow::Result;

use crate::token::Token;

#[allow(dead_code)]
pub struct Lexer {
    input: Vec<u8>,
    position: usize,
    read_position: usize,
    ch: u8,
}

#[allow(dead_code)]
impl Lexer {
    pub fn new(input: &str) -> Lexer {
        let mut lex = Lexer {
            position: 0,
            read_position: 0,
            ch: 0,
            input: input.into(),
        };
        lex.read_char();

        return lex;
    }

    fn read_ident(&mut self) -> String {
        let start = self.position;
        while self.ch.is_ascii_alphabetic() || self.ch == b'_' {
            self.read_char();
        }

        return String::from_utf8_lossy(&self.input[start..self.position]).to_string();
    }

    fn read_number(&mut self) -> Token {
        let start = self.position;
        let mut has_dot = false;
        while self.ch.is_ascii_digit() || self.ch == b'.' {
            if self.ch == b'.' && has_dot {
                return Token::Illegal;
            } else if self.ch == b'.' {
                has_dot = true
            }
            self.read_char();
        }
        let num = String::from_utf8_lossy(&self.input[start..self.position]).to_string();
        if has_dot {
            return Token::Float(num);
        }

        return Token::Int(num);
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0
        } else {
            self.ch = self.input[self.read_position]
        }
        self.position = self.read_position;
        self.read_position += 1
    }

    fn peek(&mut self) -> u8 {
        if self.read_position >= self.input.len() {
            return 0;
        } else {
            return self.input[self.read_position];
        }
    }

    pub fn next_token(&mut self) -> Result<Token> {
        self.skip_whitespace();
        let tok = match self.ch {
            b'=' => match self.peek() {
                b'=' => {
                    self.read_char();
                    Token::Eq
                }
                _ => Token::Assign,
            },
            b'-' => Token::Minus,
            b'!' => match self.peek() {
                b'=' => {
                    self.read_char();
                    Token::NotEq
                }
                _ => Token::Bang,
            },
            b'/' => Token::Slash,
            b'*' => Token::Asterisk,
            b'<' => Token::Lt,
            b'>' => Token::Gt,
            b';' => Token::Semicolon,
            b'(' => Token::Lparen,
            b')' => Token::Rparen,
            b',' => Token::Comma,
            b'+' => Token::Plus,
            b'{' => Token::Lbrace,
            b'}' => Token::Rbrace,
            b'a'..=b'z' | b'A'..=b'Z' | b'_' => {
                let ident = self.read_ident();
                return Ok(match ident.as_str() {
                    "fn" => Token::Function,
                    "let" => Token::Let,
                    "if" => Token::If,
                    "false" => Token::False,
                    "true" => Token::True,
                    "return" => Token::Return,
                    "else" => Token::Else,
                    _ => Token::Ident(ident),
                });
            }
            b'0'..=b'9' => return Ok(self.read_number()),
            0 => Token::Eof,
            _ => Token::Illegal,
        };
        self.read_char();
        Ok(tok)
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }
}

#[cfg(test)]
mod test {
    use anyhow::Result;

    use crate::lexer::Lexer;
    use crate::token::Token;

    fn get_input() -> &'static str {
        return "
let five = 5;
let ten = 10;
let two_and_half = 2.5;

let add = fn(x, y) {
x + y;
};


let fifteen = add(five, ten);
let twelve_and_half = add(ten, two_and_half);
!-/*5;
5 <10 >5;
if (5 <10) {
return true;
} else {
return false;
}
10 == 10;
10 != 9;
        ";
    }

    #[test]
    fn base_next_token() -> Result<()> {
        let input = "=+(){},;";
        let mut lexer = Lexer::new(input.into());

        let tokens = vec![
            Token::Assign,
            Token::Plus,
            Token::Lparen,
            Token::Rparen,
            Token::Lbrace,
            Token::Rbrace,
            Token::Comma,
            Token::Semicolon,
        ];

        for t in tokens {
            let token = lexer.next_token()?;
            println!("expected: {:?}, received {:?}", t, token);
            assert_eq!(t, token);
        }

        return Ok(());
    }

    #[test]
    fn text_next_token() -> Result<()> {
        let input = get_input();

        // vector of tuple of tokentype and string
        let tokens = vec![
            Token::Let,
            Token::Ident("five".to_string()),
            Token::Assign,
            Token::Int("5".to_string()),
            Token::Semicolon,
            Token::Let,
            Token::Ident("ten".to_string()),
            Token::Assign,
            Token::Int("10".to_string()),
            Token::Semicolon,
            Token::Let,
            Token::Ident("two_and_half".to_string()),
            Token::Assign,
            Token::Float("2.5".to_string()),
            Token::Semicolon,
            Token::Let,
            Token::Ident("add".to_string()),
            Token::Assign,
            Token::Function,
            Token::Lparen,
            Token::Ident("x".to_string()),
            Token::Comma,
            Token::Ident("y".to_string()),
            Token::Rparen,
            Token::Lbrace,
            Token::Ident("x".to_string()),
            Token::Plus,
            Token::Ident("y".to_string()),
            Token::Semicolon,
            Token::Rbrace,
            Token::Semicolon,
        ];

        let mut l = Lexer::new(input);

        for t in tokens {
            let token = l.next_token()?;
            println!("expected: {:?}, received {:?}", t, token);
            assert_eq!(t, token)
        }

        return Ok(());
    }
}
